nibabies/utils/bids.py

```diff
@@ -304,3 +304,51 @@ def validate_input_dir(exec_env, bids_dir, participant_label):
             subprocess.check_call(["bids-validator", str(bids_dir), "-c", temp.name])
         except FileNotFoundError:
             print("bids-validator does not appear to be installed", file=sys.stderr)
+
+
+def collect_precomputed_derivatives(layout, subject_id, session_id=None):
+    """
+    Query and collect precomputed derivatives.
+
+    This function is used to determine which workflow steps can be skipped,
+    based on the files found.
+    """
+
+    deriv_queries = {
+        'anat_mask': {
+            'datatype': 'anat',
+            'desc': 'brain',
+            'space': 'orig',
+            'suffix': 'mask',
+        },
+        'anat_aseg': {
+            'datatype': 'anat',
+            'desc': 'aseg',
+            'space': 'orig',
+            'suffix': 'dseg',
+        },
+        'bold_mask': {
+            'datatype': 'func',
+            'desc': 'brain',
+            'space': 'orig',
+            'suffix': 'mask',
+        },
+    }
+    derivatives = {}
+
+    for deriv, query in deriv_queries.items():
+        res = layout.get(
+            scope='derivatives',
+            subject=subject_id,
+            session=session_id,
+            extension=['nii', 'nii.gz'],
+            **query,
+        )
+        if not res:
+            continue
+        if len(res) > 1:
+            raise Exception(
+                f"When searching for <{deriv}>, found multiple results: {[f.path for f in res]}"
+            )
```

Comment:
For future consideration: One thing I was thinking about is we could add an option to specify entities for querying that could augment/override our queries: `--deriv-spec anat_mask=desc-nnunet anat_seg=desc-manual`

Reply:
yeah, i was thinking this during the implementation. i was considering using a config file that included rules for each derivative, in a similar vein to what you wrote:

`--derivatives-spec specs.{json/yml}`